File: graphqlclient/graphqlClient.go
package graphqlclient

import (
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/machinebox/graphql"
)

var testCounts = []int{100, 200, 300}

const baseURL = "http://127.0.0.1:5002/graphql"

var queries = map[string]string{
	"usuarios":  `{ usuarios { ID nome } }`,
	"musicas":   `{ musicas { ID nome } }`,
	"playlists": `{ playlists { ID nome } }`,
}

func testLoad(client *graphql.Client, query, name string, numTest int) {
	start := time.Now()

	var wg sync.WaitGroup

	for i := 0; i < numTest; i++ {
		wg.Add(1)
		func(i int) {
			defer wg.Done()
			req := graphql.NewRequest(query)
			ctx := context.Background()
			var res map[string]interface{}
			if err := client.Run(ctx, req, &res); err != nil {
				fmt.Printf("Request %d failed: %v\n", i, err)
				return
			}
		}(i)
	}
	wg.Wait()

	end := time.Now()
	fmt.Printf("%s load test with %d requests completed in %v seconds\n", name, numTest, end.Sub(start).Seconds())
}

func RunTests() {
	client := graphql.NewClient(baseURL)
	for _, numTest := range testCounts {
		fmt.Printf("Running tests with %d requests:\n", numTest)
		testLoad(client, queries["usuarios"], "Usuarios", numTest)
		testLoad(client, queries["musicas"], "Musicas", numTest)
		testLoad(client, queries["playlists"], "Playlists", numTest)
	}
}

File: python/clients/graphql_client.py
import requests
import time

BASE_URL = "http://127.0.0.1:5002/graphql"
TEST_COUNTS = [100, 200, 300]
QUERIES = {
    "usuarios": "{ usuarios { ID nome } }",
    "musicas": "{ musicas { ID nome } }",
    "playlists": "{ playlists { ID nome } }",
}


def test_load(query, name, num_test):
    start = time.time()

    for i in range(num_test):
        try:
            response = requests.post(BASE_URL, json={"query": query})
            response.raise_for_status()
        except requests.RequestException as e:
            print(f"Request {i} failed: {e}")

    end = time.time()
    print(
        f"{name} load test with {num_test} requests completed in {end - start:.6f} seconds"
    )


def run_tests():
    for num_test in TEST_COUNTS:
        print(f"Running tests with {num_test} requests:")
        for name, query in QUERIES.items():
            test_load(query, name, num_test)


if __name__ == "__main__":
    run_tests()
